File: Java Leetcode/Solution25.java
// 25. Reverse Nodes in k-Group

class ListNode{
    int val;
    public ListNode next;
    ListNode() {}
    ListNode(int val) {this.val = val;}
    ListNode(int val, ListNode next) {this.val = val; this.next = next;}
}

class Solution25{
    public ListNode reverseKGroup(ListNode head, int k){
        if(head == null || k == 1){
            return head;
        }

        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode ptr = dummy;

        while(ptr != null){
            ListNode track = ptr;

            for(int i = 0; i < k && track != null; i++){
                track = track.next;
            }
            if(track == null){
                break;
            }
            ListNode[] updateNode = reverseLinkedList(ptr.next, k);

            ListNode prev = updateNode[0];
            ListNode curr = updateNode[1];

            ListNode lastNodeRevGrp =ptr.next;
            lastNodeRevGrp.next = curr;
            ptr.next = prev;
            ptr = lastNodeRevGrp;
        }
        return dummy.next;

    }
    private ListNode[] reverseLinkedList(ListNode head, int k){
        ListNode prev = null;
        ListNode curr = head;
        ListNode next_node;

        for(int i = 0; i < k; i++){
            next_node = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next_node;
        }
        return new ListNode[] {prev, curr};
    }
}

File: Python Leetcode/Solution25.py
# 25. Reverse Nodes in k-Group

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head, k):
        if head is None or k == 1:
            return head

        dummy = ListNode(0)
        dummy.next = head
        ptr = dummy

        while ptr is not None:
            track = ptr

            for i in range(k):
                track = track.next
                if track is None:
                    return dummy.next

            prev, curr = self.reverseLinkedList(ptr.next, k)

            lastNodeRevGrp = ptr.next
            lastNodeRevGrp.next = curr
            ptr.next = prev
            ptr = lastNodeRevGrp

        return dummy.next

    def reverseLinkedList(self, head, k):
        prev = None
        curr = head

        for _ in range(k):
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node

        return prev, curr

